File: src/main/java/recipestore/input/RecipeResourceFilter.java
package recipestore.input;

import org.apache.jena.rdf.model.Property;
import org.apache.jena.rdf.model.Resource;
import org.apache.jena.rdf.model.StmtIterator;

import java.util.function.Predicate;

public class RecipeResourceFilter {

    public static Predicate<Resource> getRecipeWithMinimumNumberOfRating
            = (resource) -> {
        try {
            final StmtIterator ratingStmts = resource.listProperties(getAggregateRatingProperty(resource));
            if (ratingStmts.hasNext()) {
                final StmtIterator ratingValueStmt = ratingStmts.next().getObject()
                        .asResource().listProperties(getNumberOfReviewsProperty(resource));

                return (ratingValueStmt.hasNext() ? ratingValueStmt.next().getObject().asLiteral().getDouble()
                        : 0) >= 10;

            }
        } catch (Exception e) {
            return false;
        }
        return false;
    };

    private static Property getAggregateRatingProperty(Resource resource) {
        return resource.getModel().getProperty("http://schema.org/Recipe/aggregateRating");
    }

    private static Property getNumberOfReviewsProperty(Resource resource) {
        return resource.getModel().getProperty("http://schema.org/AggregateRating/reviewCount");
    }


}

File: src/main/scala/recipestore/input/RecipeResourceFilter.scala
package recipestore.input

import org.apache.jena.rdf.model.{Property, Resource}

object RecipeResourceFilter {
  def getRecipeWithMinimumNumberOfRating(resource: Resource): Boolean = {
    try {
      val ratingStmts = resource.listProperties(getAggregateRatingProperty(resource))
      if (ratingStmts.hasNext()) {
        val ratingValueStmt = ratingStmts.next().getObject()
          .asResource().listProperties(getNumberOfReviewsProperty(resource))

        (if (ratingValueStmt.hasNext()) ratingValueStmt.next().getObject().asLiteral().getDouble()
        else
          0) > 10

      }
    } catch {
      case _: Throwable => false
    }
    return false
  }

  private def getAggregateRatingProperty(resource: Resource): Property = {
    return resource.getModel.getProperty("http://schema.org/Recipe/aggregateRating")
  }

  private def getNumberOfReviewsProperty(resource: Resource): Property = {
    return resource.getModel.getProperty("http://schema.org/AggregateRating/reviewCount")
  }
}